Dockerfile

```diff
@@ -0,0 +1,105 @@
+FROM debian:stable-slim as go-builder
+# defined from build kit
+# DOCKER_BUILDKIT=1 docker build . -t ...
+ARG TARGETARCH
+
+FROM debian:stable-slim as builder
+# defined from build kit
+# DOCKER_BUILDKIT=1 docker build . -t ...
+ARG TARGETARCH
+
+RUN export DEBIAN_FRONTEND=noninteractive && \
+    apt update && \
+    apt install -y -q --no-install-recommends \
+    git curl gnupg2 build-essential \
+    linux-headers-${TARGETARCH} libc6-dev \ 
+    openssl libssl-dev pkg-config \
+    ca-certificates apt-transport-https \
+    python3 && \
+    apt clean && \
+    rm -rf /var/lib/apt/lists/*
+
+RUN useradd --create-home -s /bin/bash xmtp
+RUN usermod -a -G sudo xmtp
+RUN echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
+
+WORKDIR /rustup
+## Rust
+ADD https://sh.rustup.rs /rustup/rustup.sh
+RUN chmod 755 /rustup/rustup.sh
+
+ENV USER=xmtp
+USER xmtp
+RUN /rustup/rustup.sh -y --default-toolchain stable --profile minimal
+
+ENV PATH=$PATH:~xmtp/.cargo/bin
+
+FROM debian:stable-slim
+ARG TARGETARCH
+
+RUN export DEBIAN_FRONTEND=noninteractive && \
+  apt update && \
+  apt install -y -q --no-install-recommends \
+    ca-certificates apt-transport-https \
+    sudo ripgrep procps build-essential \
+    python3 python3-pip python3-dev \
+    git curl && \
+  apt clean && \
+  rm -rf /var/lib/apt/lists/*
+
+RUN echo "building platform $(uname -m)"
+
+RUN useradd --create-home -s /bin/bash xmtp
+RUN usermod -a -G sudo xmtp
+RUN echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
+
+## Node and NPM
+RUN mkdir -p /usr/local/nvm
+ENV NVM_DIR=/usr/local/nvm
+
+ENV NODE_VERSION=v20.9.0
+
+ADD https://raw.githubusercontent.com/creationix/nvm/master/install.sh /usr/local/etc/nvm/install.sh
+RUN bash /usr/local/etc/nvm/install.sh && \
+    bash -c ". $NVM_DIR/nvm.sh && nvm install $NODE_VERSION && nvm alias default $NODE_VERSION && nvm use default"
+
+ENV NVM_NODE_PATH ${NVM_DIR}/versions/node/${NODE_VERSION}
+ENV NODE_PATH ${NVM_NODE_PATH}/lib/node_modules
+ENV PATH      ${NVM_NODE_PATH}/bin:$PATH
+
+RUN npm install npm -g
+RUN npm install yarn -g
+
+
+## Rust from builder
+COPY --chown=xmtp:xmtp --from=builder /home/xmtp/.cargo /home/xmtp/.cargo
+COPY --chown=xmtp:xmtp --from=builder /home/xmtp/.rustup /home/xmtp/.rustup
+
+USER xmtp
+
+# fmt coming from nightly
+RUN ~xmtp/.cargo/bin/rustup toolchain install nightly
```

Comment:
One thing that's funky here is that the Docker caching is going to lock you to whichever nightly version you first built this with. Subsequent builds will not re-download a newer nightly version.

If the goal is reproducibility, we should really have this locked to a particular stable version and make updates explicit.

Reply:
It looks like @insipx created a ticket to remove the nightly.   I agree I think stable seems like the route to take.

Reply:
In the fmt nightly ticket, i'll add to lock to a particular stable version. I think there should be a way with rustup to point it to the rust-toolchain file and install all the necessary components through that